pydatastructs/strings/suffix_tree.py

```diff
@@ -0,0 +1,273 @@
+                return 'list'
+        raise ValueError("String argument should be of type String or a list of strings")
+
+    def build(self, x):
+        """
+        Builds the Suffix tree on the given input.
+
+        Parameters
+        ==========
+
+        x: str or list of str
+
+        Returns
+        =======
+
+        None
+        """
+        type = self._check_input(x)
+        if type == 'str':
+            x += next(self._terminalSymbolsGenerator())
+            self._build(x)
+        if type == 'list':
+            self._build_generalized(x)
+
+    def _build(self, x):
+        self.word = x
+        self._build_McCreight(x)
+
+    def _build_McCreight(self, x):
+        u = self.root
+        d = 0
+        for i in range(len(x)):
+            while u.depth == d and u._has_transition(x[d + i]):
+                u = u._get_transition_link(x[d + i])
+                d = d + 1
+                while d < u.depth and x[u.idx + d] == x[i + d]:
+                    d = d + 1
+            if d < u.depth:
+                u = self._create_node(x, u, d)
+            self._create_leaf(x, i, u, d)
+            if not u._get_suffix_link():
+                self._compute_slink(x, u)
+            u = u._get_suffix_link()
+            d = d - 1
+            if d < 0:
+                d = 0
+
+    def _create_node(self, x, u, d):
+        i = u.idx
+        p = u.parent
+        v = SuffixNode(idx=i, depth=d)
+        v._add_transition_link(u, x[i + d])
+        u.parent = v
+        p._add_transition_link(v, x[i + p.depth])
+        v.parent = p
+        return v
+
+    def _create_leaf(self, x, i, u, d):
+        w = SuffixNode()
+        w.idx = i
+        w.depth = len(x) - i
+        u._add_transition_link(w, x[i + d])
+        w.parent = u
+        return w
+
+    def _compute_slink(self, x, u):
+        d = u.depth
+        v = u.parent._get_suffix_link()
+        while v.depth < d - 1:
+            v = v._get_transition_link(x[u.idx + v.depth + 1])
+        if v.depth > d - 1:
+            v = self._create_node(x, v, d - 1)
+        u._add_suffix_link(v)
+
+    def _build_generalized(self, xs):
+        terminal_gen = self._terminalSymbolsGenerator()
+        _xs = ''.join([x + next(terminal_gen) for x in xs])
+        self.word = _xs
+        self._generalized_word_starts(xs)
+        self._build(_xs)
+        self.root._traverse(self._label_generalized)
+
+    def _label_generalized(self, node):
+        if node.is_leaf():
+            x = {self._get_word_start_index(node.idx)}
+        else:
+            x = {n for ns in node.transition_links.values() for n in ns.generalized_idxs}
+        node.generalized_idxs = x
+
+    def _get_word_start_index(self, idx):
+        i = 0
+        for _idx in self.word_starts[1:]:
+            if idx < _idx:
+                return i
+            else:
+                i += 1
+        return i
+
+    def longest_common_substring(self, stringIdxs = -1):
```

Comment:
What if I want to pass a string as a query input?

Reply:
LCS finds the common substring among the given list of strings as for this if a new string which has to be check must be added in the initial tree instance, do you mean the character to start check of lcs from